File: backend/schemas/finance.py
from pydantic import BaseModel


class IncomeRequest(BaseModel):
    amount: float
    currency: str = "UAH"
    original_amount: float | None = None


class ExpenseRequest(BaseModel):
    category: str
    amount: float
    currency: str = "UAH"


class SyncRequest(BaseModel):
    target_total: float


class BalancesResponse(BaseModel):
    mandatory: float
    non_mandatory: float
    investments: float
    dreams: float


class ExpenseResponse(BaseModel):
    balances: BalancesResponse
    breach_note: str | None


class FlushResponse(BaseModel):
    flushed_total: float
    balances: BalancesResponse


class RateResponse(BaseModel):
    currency: str
    rate: float


# --- TASK-005 schemas ---


class TransactionResponse(BaseModel):
    id: str
    date: str
    type: str
    category: str
    amount_uah: float
    original_amount: float | None
    original_currency: str | None
    envelope: str
    details: str


class StatsEnvelope(BaseModel):
    total: float
    categories: dict[str, float]


class StatsResponse(BaseModel):
    mandatory: StatsEnvelope
    non_mandatory: StatsEnvelope
    investments: StatsEnvelope
    dreams: StatsEnvelope


class AuditResponse(BaseModel):
    breach_count: int
    breach_total_uah: float
    breach_by_envelope: dict[str, float]
    top_breaches: list[dict]
    total_spent_uah: float
    burn_rate_daily: float
    days_remaining: int
    days_to_zero: float | None
    safe_daily_limit: float
    spendable_balance: float
    health_signal: str


class SustainabilityResponse(BaseModel):
    mandatory_burn: float
    non_mandatory_burn: float
    combined_burn: float
    days_to_zero_mandatory: float | None
    days_to_zero_non_mandatory: float | None
    days_to_zero_combined: float | None
    safe_daily_mandatory: float
    safe_daily_non_mandatory: float
    safe_daily_combined: float
    days_remaining: int
    mandatory_spent: float
    non_mandatory_spent: float


class AnomalyItem(BaseModel):
    category: str
    last_7d: float
    avg_7d: float
    ratio: float


class AdvisorResponse(BaseModel):
    text: str


class ImpactRequest(BaseModel):
    amount: float
    category: str | None = None


class ImpactResponse(BaseModel):
    amount: float
    category: str | None
    spendable_before: float
    spendable_after: float
    daily_limit_before: float
    daily_limit_after: float
    days_remaining: int
    waterfall_triggered: bool
    risk_score: str
    category_valid: bool


class CategoryRequest(BaseModel):
    name: str
    envelope_name: str


class CategoryResponse(BaseModel):
    name: str
    envelope_name: str


class ConfigRequest(BaseModel):
    base_currency: str


class ConfigResponse(BaseModel):
    base_currency: str

File: mobile/src/types/finance.ts
// Request types

export interface IncomeRequest {
  amount: number;
  currency?: string;
  original_amount?: number | null;
}

export interface ExpenseRequest {
  category: string;
  amount: number;
  currency?: string;
}

export interface SyncRequest {
  target_total: number;
}

export interface ImpactRequest {
  amount: number;
  category?: string | null;
}

export interface CategoryRequest {
  name: string;
  envelope_name: string;
}

export interface ConfigRequest {
  base_currency: string;
}

// Response types

export interface BalancesResponse {
  mandatory: number;
  non_mandatory: number;
  investments: number;
  dreams: number;
}

export interface ExpenseResponse {
  balances: BalancesResponse;
  breach_note: string | null;
}

export interface FlushResponse {
  flushed_total: number;
  balances: BalancesResponse;
}

export interface RateResponse {
  currency: string;
  rate: number;
}

export interface TransactionResponse {
  id: string;
  date: string;
  type: string;
  category: string;
  amount_uah: number;
  original_amount: number | null;
  original_currency: string | null;
  envelope: string;
  details: string;
}

export interface StatsEnvelope {
  total: number;
  categories: Record<string, number>;
}

export interface StatsResponse {
  mandatory: StatsEnvelope;
  non_mandatory: StatsEnvelope;
  investments: StatsEnvelope;
  dreams: StatsEnvelope;
}

export interface AuditResponse {
  breach_count: number;
  breach_total_uah: number;
  breach_by_envelope: Record<string, number>;
  top_breaches: Record<string, unknown>[];
  total_spent_uah: number;
  burn_rate_daily: number;
  days_remaining: number;
  days_to_zero: number | null;
  safe_daily_limit: number;
  spendable_balance: number;
  health_signal: string;
}

export interface SustainabilityResponse {
  mandatory_burn: number;
  non_mandatory_burn: number;
  combined_burn: number;
  days_to_zero_mandatory: number | null;
  days_to_zero_non_mandatory: number | null;
  days_to_zero_combined: number | null;
  safe_daily_mandatory: number;
  safe_daily_non_mandatory: number;
  safe_daily_combined: number;
  days_remaining: number;
  mandatory_spent: number;
  non_mandatory_spent: number;
}

export interface AnomalyItem {
  category: string;
  last_7d: number;
  avg_7d: number;
  ratio: number;
}

export interface AdvisorResponse {
  text: string;
}

export interface ImpactResponse {
  amount: number;
  category: string | null;
  spendable_before: number;
  spendable_after: number;
  daily_limit_before: number;
  daily_limit_after: number;
  days_remaining: number;
  waterfall_triggered: boolean;
  risk_score: string;
  category_valid: boolean;
}

export interface CategoryResponse {
  name: string;
  envelope_name: string;
}

export interface ConfigResponse {
  base_currency: string;
}

// Pending writes (offline queue)

export interface PendingWrite {
  id: string;
  operation: 'addExpense' | 'addIncome' | 'removeTransaction';
  payload: string;
  created_at: string;
  attempts: number;
  last_error: string | null;
  status: 'pending' | 'syncing' | 'failed';
}

// Auth types

export interface AuthTokens {
  access_token: string;
  token_type: string;
}

export interface RegisterRequest {
  email: string;
  password: string;
}

export interface LoginRequest {
  email: string;
  password: string;
}
